### DISCUSSION TITLE
Input validation with Prisma?

### DISCUSSION THREAD

**Discussion by IHIutch** [trpc/trpc#3707] [General]
Hello all. Jsyk, I'm fairly new to typescript (tRPC got me to start learning!) so I apologize if this is a silly question.

Anyway, I've been playing with tRPC, zod, and Prisma types and that made me wonder if there is a way to use the types generated by prisma to validate inputs. 

For example, here is a setup with zod, but I've noticed things like find, create, update can get pretty complex (like if I want to be able to reuse my block schema within those actions. In this case I might need to set some or all of my parameter as optional (or use `pick` or `omit`... etc)).
```
const PostSchema = z.object({
  id: z.number(),
  title: z.string(),
  content: z.string(),
})

export const appRouter = router({
  byId: publicProcedure
    .input(PostSchema.optional())
    .query(({ input }) => {
      return {
        data: await getPost({id})
      };
    }),
});
```

But if you're using Prisma, those different types are generated for you. And since the types are generated from your DB schema it might be nice to leverage those types without having to manage your zod types as well. So it'd be nice to be able to do something like this:
```
const PostSchema = Prisma.PostWhereUniqueInput

// type Prisma.PostWhereUniqueInput = {
//     id?: string | undefined;
//     title?: string | undefined;
//     content?: string | undefined;
// }

export const appRouter = router({
  byId: publicProcedure
    .input(PostSchema)
    .query(({ input }) => {
      return {
        data: await getPost({id})
      };
    }),
});
```

**Comment by sachinraja** [comment]
https://github.com/CarterGrimmeisen/zod-prisma

**Reply by IHIutch** [reply]
I saw that, and I was looking through this repo too  https://github.com/omar-dulaimi/prisma-trpc-generator


They seem like fine workarounds, but they still require Prisma and Zod to be/stay in sync.

I can see the utility of using both Zod and Prisma types on a project, but specifically at the input validation step it seems like Prisma types would be the most useful (assuming the next step tends to be a write/read from the DB. At least, that's the way my routers are set up)

**Comment by IHIutch** [comment]
After continuing to read and look into this a bit more, I'm thinking this idea might be a misunderstanding of purely "typing" (Prisma) and "validation" (Zod). I recognize that Zod can handle typing as well, but maybe that's part of the reason these are separate? 

I'd greatly appreciate if anyone could elaborate here, if I'm thinking about this correctly or if I'm way off base.

**Reply by medv** [reply]
This is correct. Prisma is types for lower level database operations. Zod is complex runtime validation (much more than typescript types can do, besides they are stripped at compile time). However, Zod also provides typescript types that describe its schemas to the best of its ability.

In my opinion, you would use Zod as the front-facing contract wrapping (but not completely replicating) prisma's methods. Your public facing api should not expose all of prisma's ablities, unless you are making a CMS (and even then, you can still simplify it to just the operations and arguments you actually need in your CMS), then pass the zod-validated objects to the underlying prisma operations.
